xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc

```diff
@@ -526,6 +526,338 @@ class CholeskyOpLowering : public OpRewritePattern<CholeskyOp> {
+
+class FusedAttentionForwardOpLowering
+    : public FusedAttentionForwardLowering<fusedMHAOp> {
+ public:
+  using FusedAttentionForwardLowering::FusedAttentionForwardLowering;
+};
+
+class FusedAttentionScaledMaskForwardOpLowering
+    : public FusedAttentionForwardLowering<fusedMHAWithScaledMaskOp> {
+ public:
+  using FusedAttentionForwardLowering::FusedAttentionForwardLowering;
+};
+
+class FusedAttentionScaledBiasForwardOpLowering
+    : public FusedAttentionForwardLowering<fusedMHAWithScaledBiasOp> {
+ public:
+  using FusedAttentionForwardLowering::FusedAttentionForwardLowering;
+};
+
+using mlir::lmhlo_gpu::fusedMHABackwardOp;
+using mlir::lmhlo_gpu::fusedMHAWithMaskBackwardOp;
+
+template <typename FusedDotAttentionBackward>
+class FusedAttentionBackwardLowering
+    : public OpRewritePattern<FusedDotAttentionBackward> {
+ private:
+  static constexpr const char kCustomCallTarget[] =
+      "xla.gpu.fused.attention.backward.";
+
+ public:
+  explicit FusedAttentionBackwardLowering(MLIRContext* ctx, UidGenerator& uid,
+                                          CustomCallDeclarations& custom_calls)
+      : OpRewritePattern<FusedDotAttentionBackward>(ctx),
+        uid_(uid),
+        custom_calls_(custom_calls) {}
+
+  LogicalResult matchAndRewrite(FusedDotAttentionBackward op,
+                                PatternRewriter& rewriter) const override {
+    // if (auto bwd_op =
+    // dyn_cast<fusedMHAWithMaskBackwardOp>(op.getOperation())) {
+    //   return op.emitOpError("fusedMHAWithMaskBackwardOp not supported yet.");
+    // }
+    // Get the custom call target.
+    std::string fused_attention = kCustomCallTarget;
+    auto num_operands = op.getNumOperands();
+    switch (op.getFusedMhaDag()) {
+      case mlir::lmhlo_gpu::FusedMhaBackwardDagSignature::
+          BackwardScaleBiasSoftmax:
+        if (num_operands == 10) {
+          fused_attention += "scale.softmax";
+        } else if (num_operands == 11) {
+          fused_attention += "scale.dbias.softmax";
+        } else
+          return op.emitOpError(
+              "unexpected number of operands for fused attention backward - "
+              "BMM_Bias_Softmax_BMM");
+        break;
+
+      case mlir::lmhlo_gpu::FusedMhaBackwardDagSignature::
+          BackwardScaleBiasSoftmaxDropout:
+        if (num_operands == 10) {
+          fused_attention += "scale.softmax.dropout";
+        } else if (num_operands == 11) {
+          fused_attention += "scale.dbias.softmax.dropout";
+        } else
+          return op.emitOpError(
+              "unexpected number of operands for fused attention backward - "
+              "BMM_Bias_Softmax_Dropout_BMM");
+        break;
+
+      case mlir::lmhlo_gpu::FusedMhaBackwardDagSignature::
+          BackwardScaleBiasMaskSoftmax:
+        if (num_operands == 11) {
+          fused_attention += "scale.mask.softmax";
+        } else if (num_operands == 12) {
+          fused_attention += "scale.dbias.mask.softmax";
+        } else
```

Comment:
add braces to all `else` cases in this file, it should be consistent between all branches in the if statement